jlama-core/src/main/java/com/github/tjake/jlama/model/PromptBuilder.java

```
@@ -0,0 +1,90 @@
+
+  PromptBuilder(AbstractModel model) {
+    this.model = model;
+  }
+
+  public PromptBuilder addUserMessage(String userMessage) {
+    this.messages.computeIfAbsent("user", k -> new ArrayList<>()).add(userMessage);
+    return this;
+  }
+
+  public PromptBuilder addSystemMessage(String systemMessage) {
+    this.messages.computeIfAbsent("system", k -> new ArrayList<>()).add(systemMessage);
+    return this;
+  }
+
+  public PromptBuilder addAssistantMessage(String assistantMessage) {
+    this.messages.computeIfAbsent("assistant", k -> new ArrayList<>()).add(assistantMessage);
+    return this;
+  }
+
+  public PromptBuilder addToolCall(ToolCall toolCall) {
+    this.toolCalls.add(toolCall);
+    return this;
+  }
+
+  public PromptBuilder addToolResult(ToolResult toolResult) {
+    this.toolResults.add(toolResult);
+    return this;
+  }
+
+  public PromptBuilder generationalPrompt(boolean isGenerational) {
+    this.isGenerational = isGenerational;
+    return this;
+  }
+
+  public PromptContext build() {
+    PromptContext ctx;
+
+    if (model.promptSupport().isPresent()) {
+      var promptSupport = model.promptSupport().get().builder();
+
+      if (this.messages.containsKey("user")) {
+        this.messages.get("user").forEach(promptSupport::addUserMessage);
+      }
+
+      if (this.messages.containsKey("system")) {
+        this.messages.get("system").forEach(promptSupport::addSystemMessage);
+      }
+
+      if (this.messages.containsKey("assistant")) {
+        this.messages.get("assistant").forEach(promptSupport::addAssistantMessage);
+      }
+
+      if (!this.toolCalls.isEmpty()) {
+        this.toolCalls.forEach(promptSupport::addToolCall);
+      }
+
+      if (!this.toolResults.isEmpty()) {
+        this.toolResults.forEach(promptSupport::addToolResult);
+      }
+
+      promptSupport.addGenerationPrompt(isGenerational);
+
+      ctx = promptSupport.build();
+    } else {
+      var userMessages = this.messages.get("user");
```

Hmm is this how no prompt template should be handled? Or should we just concatenate all messages?  System -> User -> Assistant -> User -> Assistant... etc

---

I don't understand the question.

---

Maybe just assert that only user messages are set?  Since other message types would be lost?  Also when many user messages are specified they should be concatenated with "\n".

---

For the second case, can you throw an exception if anything other than user messages is set?

Yes "\n".

---

I think that if anything other than user message is set will be ignored, because, to generate PromptContext, user message is only what we need.

---

Ok, how about just a log message warning them its going to be ignored?

---

A warning that model does not support prompt? If is it, I agree.

---

Yes

---

A moment...

---

Done.